Typing and export PresenceError

Amund211 opened the pull request:
Just upstreaming some more of the changes I've made in my copy of your library (https://github.com/Amund211/prism/commit/358bd76b86efe83ed393b6482587d761ae742fb2), though a bit less invasive this time. Feel free to pick and choose which commits you want or don't want.

Summary of the changes:
- Made explicit that `Presence` is exported from `__init__.py` (`mypy` was complaining about this)
- Added some more type hints in `presence.py` to make it pass `mypy --strict` (there are still some `Any` types and 1 call to `cast`
- Added a `py.typed` marker. This makes the type hints visible to people using the library, as before you would get an error that the module was installed, but could not be type-checked
- Added `PresenceError` as an export from `__init__.py`
- Fix type-checking on windows
- Always do platform check the same way (ref last commit)

Amund211 committed:
Specify export in __init__.py

Amund211 committed:
Fix typing in presence.py

Amund211 committed:
Become PEP 561-compliant with a py.typed marker

Previously type checkers like mypy would give an error when typechecking
code using the library: Skipping analyzing "discordrp": module is
installed, but missing library stubs or py.typed marker
This patch fixes this issue.

Amund211 committed:
Also export PresenceError from __init__.py

Amund211 committed:
Fix sys.platform type guard

According to
https://peps.python.org/pep-0484/#version-and-platform-checking
type checkers should understand simple platform checks, however
`self._platform == WINDOWS` is not simple enough for mypy 1.3.0.
Replaced it with `sys.platform == 'win32'` so the type-checker doesn't
get mad at `socket.AF_UNIX` on windows.

Amund211 committed:
Always compare sys.platform directly with 'win32'

This doesn't help the type-checking, but makes all the platform checks
done in the same way in an attempt to be less confusing.

Amund211 commented:
For the `sys.platform` thing I tried keeping the variable `WINDOWS` so we don't have to duplicate the string literal everywhere, but `mypy` didn't like it.
https://github.com/Amund211/prism/commit/70ebb79b3c5cad55379a11046d24aafd72ebc488: https://github.com/Amund211/prism/actions/runs/5037578368/jobs/9034513970
https://github.com/Amund211/prism/commit/d918bea15204557ff692f17f5e05ee10e0b4abab: https://github.com/Amund211/prism/actions/runs/5037594511/jobs/9034540744

TenType commented:
Thanks for all the help. Right now I am occupied with work but I'll take a look at it when I get the chance.

Amund211 committed:
Fix typing of __exit__

When no exception is raised these variables are None

TenType committed:
Refactor sockets into different classes

TenType reviewed (commented):
Sorry for the delay! At this point, I think it would be better if separate socket classes were made for both platforms to avoid leaving the `socket` typed as `Any`. I kept your commits but also added my own refactoring. From my end on MacOS, it's still `mypy --strict` compliant, but let me know if it doesn't work for you or if you have any suggestions.

Amund211 committed:
typing: Replace new union syntax with Optional

The library supports python 3.9, which does not support union with |

Amund211 committed:
typing: Type ignore socket.AF_UNIX

This fixes typechecking on windows, as socket.AF_UNIX is not defined
there. Adding unused-ignore also fixes the unused ignore error on mac
and linux.

TenType commented on discordrp/presence.py:
```diff
@@ -4,14 +4,17 @@
+from abc import ABC, abstractmethod
 from enum import IntEnum
-from typing import Any
+from typing import Any, Union, cast
```
```suggestion
from typing import Any, Optional, cast
```

TenType commented on discordrp/presence.py:
```diff
@@ -34,16 +37,14 @@ class Presence:
         self._handshake()
 
-    def set(self, activity):
+    def set(self, activity: Union[dict[str, Any], None]) -> None:
```
```suggestion
    def set(self, activity: Optional[dict[str, Any]]) -> None:
```

TenType commented on discordrp/presence.py:
```diff
@@ -86,100 +87,128 @@ def set(self, activity):
-    def __exit__(self, exc_type, exc_value, exc_traceback):
+    def __exit__(
+        self,
+        exc_type: Union[type[BaseException], None],
+        exc_value: Union[BaseException, None],
+        exc_traceback: Union[TracebackType, None],
```
```suggestion
        exc_type: Optional[type[BaseException]],
        exc_value: Optional[BaseException],
        exc_traceback: Optional[TracebackType],
```

TenType reviewed (changes requested):
Prefer `Optional[T]` over `Union[T, None]`

Amund211 commented:
Separate socket classes might be the cleanest option. We're still left with errors from `mypy` on windows, but that's just because the logic that chooses the `_Socket` implementation is way to advanced for `mypy` to understand. See: https://github.com/Amund211/discord-rich-presence/actions/runs/5171453202

Adding a `# type: ignore` just gives errors on the unicies because it's not a type error there lmao. `# type: ignore [attr-defined,unused-ignore]` [should work in the next release of `mypy`](https://github.com/python/mypy/issues/8823#issuecomment-1556288015), other than that I don't really know how to fix it. Then again, I don't think client code is affected by internal type errors in a library anyway, so it might be just okay to leave it, at least until the next `mypy` version. If that sounds OK to you you can cherry pick this: https://github.com/Amund211/discord-rich-presence/commit/4e6e497ff6a9e7b83b7670aac22870eb97b847fa.

Also, when adding the action to test the typing on windows I realized that I broke python 3.9 compatibility with the new union syntax, so I changed it to the old one in 0a5623c717d0e2a29220fb6da43866045c6a2b5c (now 48f192f75e4724c3724a1e07674b46cf98ee065b).

:^)

Amund211 commented:
> Prefer `Optional[T]` over `Union[T, None]`

Fair, I'll just change the commit and force push

TenType commented:
Thanks! I'll use `# type: ignore [attr-defined,unused-ignore]` for now then.
